### PR TITLE
Lecture 2 plp and demodata and pdp

### PR DESCRIPTION
**Pull request by nika-doroshkevich**

### CODE REVIEW

File: src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java [hunk 1]
```diff
@@ -40,15 +57,61 @@ public Product getProduct(Long id) throws NoSuchElementException {
...
+    @Override
+    public List<Product> findProducts(String query, SortField sortField, SortOrder sortOrder) {
         Lock readLock = rwLock.readLock();
         readLock.lock();
 
         try {
+            var comparedByField = compareByField(sortField);
+            var comparedByFieldAndOrder = compareByOrder(comparedByField, sortOrder);
+
+            if (query == null || query.trim().isEmpty()) {
```
Comment (AnnaKrikun) [inline src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java]: use StringUtils.isNotBlank(query)

File: src/main/java/com/es/phoneshop/exception/ExceptionHandler.java [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+import jakarta.servlet.ServletException;
+import jakarta.servlet.http.HttpServlet;
+import jakarta.servlet.http.HttpServletRequest;
+import jakarta.servlet.http.HttpServletResponse;
+
+import java.io.IOException;
+import java.util.NoSuchElementException;
+
+public class ExceptionHandler extends HttpServlet {
```
Comment (AnnaKrikun) [inline src/main/java/com/es/phoneshop/exception/ExceptionHandler.java:11]: Can error pages be mapped in web.xml?

File: src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java [hunk 3]
```diff
@@ -40,15 +57,61 @@ public Product getProduct(Long id) throws NoSuchElementException {
...
+                    .sorted(Comparator.comparing(ProductDto::getNumberOfMatches).reversed())
+                    .map(ProductDto::getProduct)
+                    .sorted(comparedByFieldAndOrder)
                     .filter(Product::isAvailableForSale)
                     .collect(Collectors.toList());
+
+            //For second implementation
+//            return products.stream()
+//                    .filter(combineOr(queryWords))
+//                    .collect(Collectors.toList());
```
Comment (AnnaKrikun) [inline src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java]: What is it for??

File: src/main/java/com/es/phoneshop/web/ProductDetailsPageServlet.java [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+    private void dispatchRequest(HttpServletRequest request, HttpServletResponse response,
+                                 String requestPage, Map<String, Object> attributes) throws ServletException, IOException {
+        attributes.forEach(request::setAttribute);
+        request.getRequestDispatcher("/WEB-INF/pages/" + requestPage).forward(request, response);
+    }
+
+    private void validatePath(String path) {
+        boolean wrongPath = false;
+
+        if (path == null || path.trim().isEmpty()) {
```
Comment (AnnaKrikun) [inline src/main/java/com/es/phoneshop/web/ProductDetailsPageServlet.java]: Use StringUtils.isNotBalnk(path)

Comment (nika-doroshkevich) [reply src/main/java/com/es/phoneshop/web/ProductDetailsPageServlet.java]: Ok

File: src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java (hunk 1 above)
Comment (nika-doroshkevich) [reply src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java]: Ok, thanks

File: src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java (hunk 3 above)
Comment (nika-doroshkevich) [reply src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java]: It is a call of "combineOr" method. The "combineOr" and "filter" methods are just the second implementation of the first task (but without sorting). I just wanted to show it to you)

File: src/main/java/com/es/phoneshop/exception/ExceptionHandler.java (hunk 2 above)
Comment (nika-doroshkevich) [reply src/main/java/com/es/phoneshop/exception/ExceptionHandler.java:11]: Yep, but I needed to set the attribute "errorMessage" to display the error for NoSuchElementException. And 500 error is also here, so everything is in one place.

**Review by AnnaKrikun — APPROVED** [review]

**Pull request merged**